parsely/src/main/java/com/parsely/parselyandroid/ParselyTracker.java

```diff
@@ -484,17 +495,12 @@ private void enqueueEvent(Map<String, Object> event) {
     }
 
     /**
-     * Flush events to Parsely.
-     * <p>
-     * Empties the event queue and sends the appropriate requests to Parsely.
-     * Called automatically after a number of seconds determined by {@link #getFlushInterval()}.
-     * <p>
-     * To make sure all of the queued events are flushed to Parse.ly's servers,
-     * call this method in your main activity's `onDestroy()` method.
+     * Deprecated since 3.1.1. The SDK now automatically flushes the queue on app lifecycle events.
+     * Any usage of this method is safe to remove and will have no effect. Keeping for backwards compatibility.
      */
+    @Deprecated
     public void flushEventQueue() {
-        // needed for call from MainActivity
-        new FlushQueue().execute();
+        // no-op
```

Comment:
Question (❓): Do we really want to do that here, removing the `flushEventQueue()` implementation, while at the same time keeping the public API, let me explain... 🤔

When apps will start updating to this new version of Parsely, they might still depend on flushing the event queue manually using this public API, and not only on during `onDestroy(...)`, just like we do with our example app. Also, there are valid cases when `onDestroy(...)` is to be called when the user is still using the app, before putting it to background, like finishing at activity, etc, right? Apps still depending on this public but deprecated API might not think to check the implementation and realize that it is now a `no-op`. Thus, they will still think that the `flushEventQueue` is working as expected, which might create confusion, or even worse, potential problems for some apps, no? 🤔

As such, I recommend keeping the implementation of this public API, along with its deleted comment, and just added the `@Deprecated` annotation with its extra comment. It will not harm to have potentially 2 simultaneous calls to flush the event queue, wouldn't it, not sure, but I remember you creating some mechanism that checks if the flush is active to not trigger another flush, right? 🤔 🤔 

I also recommend keeping the `Flush Event Queue` button on the example app and use that when we need to flush the event queue manual. I think it might be better to do that rather than having engineers to know to background the app in order to achieve the same result, wdyt? 🤔 🤔 🤔

PS: We could also remove the public API along with its implementation altogether, but this feels a bit risky and disruptive at this point, maybe we should experiment with the lifecycle event observer first, see how and whether is work 100% as we expect and then move forward with removing this deprecated API.

Reply:
> they might still depend on flushing the event queue manually using this public API, and not only on during onDestroy(...), just like we do with our example app

I don't think it's the case. The only use case for `flushEventQueue` was to ask consumers of the library to add it to app/activity lifecycle, to assert sending events if app goes to background/gets destroyed. It's explained in documentation, and there's no business value in sending a set of events to Parsely at a specific time - this is the responsibility of Parsely SDK itself.

> It will not harm to have potentially 2 simultaneous calls to flush the event queue, wouldn't it, not sure, but I remember you creating some mechanism that checks if the flush is active to not trigger another flush, right? 🤔 🤔

At this moment, it would potentially harm implementation (we didn't introduce #66 ). Even if we introduce the fix, it'll make confusion on how to use API. If events are sent automatically, why bother with `flushEventQueue` manually?

> We could also remove the public API along with its implementation altogether,

Yeah, I wouldn't vote for this. Making the method no-op and deprecating is has no negative consequences, and it doesn't introduce a breaking change for the clients.

----

Also, it's worth to mention that [iOS SDK](https://docs.parse.ly/ios-sdk/) doesn't have any method that allows consumers to flush events queue and it works fine for them.

Reply:
I forgot to mention that putting `flushEventQueue` in `onDestroy`, as documentation suggests, has no effect on modern Android devices. The `onDestroy` happens so quickly that AsyncTask run in `flushEventQueue` doesn't have enough time to send POST request.

Reply:
👋 @wzieba and thanks for the reply, you've convinced me, please proceed with this change as you intended. Maybe I am just being too cautious with this, and not having it all in my head, just like you do! 🍀 

> we didn't introduce https://github.com/Parsely/parsely-android/pull/66

Ah, I was looking for that, I thought that this got merged for some reason, but couldn't find it, now I know why, thanks... 😅 

> Also, it's worth to mention that [iOS SDK](https://docs.parse.ly/ios-sdk/) doesn't have any method that allows consumers to flush events queue and it works fine for them.

Good to know! 💯

Reply:
Thanks @ParaskP7 ! This was a completely valid concern, and I'm glad we documented it for the future and any 3rd party developers to see motivations behind this change!